elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -79,6 +118,157 @@ private void logContext(RelationshipOp operation, Environment environment) {
+                               GraphQLType outputType, EntityDictionary dictionary,
+                               Optional<String> sort, Optional<String> offset,
+                               Optional<String> first, Optional<String> filters) {
+        /* fetching a collection */
+        if(!ids.isPresent()) {
+            List<PersistentResource> records = new ArrayList<>(PersistentResource.
+                    loadRecords(recordType, requestScope, Optional.empty()));
+
+            //TODO: paginate/filter/sort
+            return records;
+
+        } else { /* fetching by id(s) */
+            List<String> idList = ids.get();
+
+            /* handle empty list of ids */
+            if(idList.isEmpty()) {
+                throw new IllegalArgumentException("Empty list passed to ids");
+            }
+
+            /* access records from internal db and return */
+            Optional<FilterExpression> filterExpression;
+            GraphQLObjectType graphQLType = (GraphQLObjectType) ((GraphQLList) outputType).getWrappedType();
+            String entityType = graphQLType.getName();
+            Class entityClass = dictionary.getEntityClass(entityType);
+
+            Class<?> idType = dictionary.getIdType(entityClass);
+
+            String idField = dictionary.getIdFieldName(entityClass);
+
+            String relation = dictionary.getJsonAliasFor(entityClass);
+
+            /* construct a new SQL like filter expression, eg: book.id IN [1,2] */
+            filterExpression = Optional.of(new FilterPredicate(
+                    new FilterPredicate.PathElement(
+                            entityClass,
+                            relation,
+                            idType,
+                            idField),
+                    Operator.IN,
+                    new ArrayList<>(idList)));
+
+            return PersistentResource.loadRecords(recordType, requestScope, filterExpression);
+        }
+    }
+
+    /**
+     * FETCH attributes of top level entity
+     * @param parentResource parent object
+     * @param field Field type
+     * @return list of persistent resource objects
+     */
+    private Object fetchAttribute(PersistentResource parentResource, Field field) {
+        return parentResource.getAttribute(field.getName());
+    }
+
+    /**
+     * Fetches a relationship for a top-level entity
+     * @param context environment encapsulating graphQL's request environment
+     * @return persistence resource object(s)
+     */
+    private Object fetchRelationship(Environment context) {
+        EntityDictionary dictionary = context.requestScope.getDictionary();
+        Class parentClass = context.parentResource.getResourceClass();
+        String fieldName = context.field.getName();
+
+        /* check for toOne relationships */
+        if(dictionary.getRelationshipType(parentClass, fieldName).isToOne()) {
+            DataStoreTransaction tx = context.requestScope.getTransaction();
```

Comment:
You need to go through the persistent resource (instead of the data store), to ensure security is invoked.